text/0000-mark-cf.md

```diff
@@ -0,0 +1,159 @@
+# Mark CF
+
+- RFC PR: (to be filled)
+- Tracking issue: (to be filled)
+
+## Summary
+
+Add a new CF (column family) `mark` to store `Lock` and `Rollback` records of transactions. Whenever a `Lock` or `Rollback` record is written to the write CF, also write a copy to the mark CF. Then, the `Lock` and `Rollback` records can be removed once they are not the latest version, so they won't affect the read performance a lot. 
+
+## Motivation
+
+Consider a use case: the user application always reads and locks a fixed key but never changes it:
+
+```
+txn = client.start_transaction()
+txn.lock("k1")
+v1 = txn.get("k1")
+txn.put("k2", "v2")
+txn.commit()
+```
+
+No matter it's an optimistic or pessimistic transaction, the transaction will write a new `Lock` record on `k1` in the write CF. If we repeats executing this transaction, there can be a lot of `Lock` records on `k1` that are newer than the effective `PUT` record, like:
+
+```
+k1@100  Lock
+k1@99   Lock
+k1@98   Lock
+...
+k1@11   Lock
+k1@10   PUT   value = "v1"
+```
+
+It will be a disaster when we want to read the value of `k1`. If it's a point get, we can only seek to the latest version and waste a lot of effort calling `next` until we find an effective record. The performance of range scan is also degraded. Furthermore, the _current_ MVCC GC cannot remove these `Lock` records because they are newer than the latest `PUT` records. So, the performance will keep getting worse until the key is effectively modified.
+
+## Design
+
+`Lock` and `Rollback` records have no value in them. They are totally ignored during reading. If they are causing much trouble, why can't we remove them? Let's find out what role they perform now.
+
+### Background
+
+`Rollback` records exist to ensure the corresponding transaction must not succeed in committing. When acquiring a pessimistic lock of a key or prewriting a key, if a `Rollback` record _with the same `start_ts`_ exist on the key, it must not succeed.
+
+We already have a collapsing mechanism to merge consecutive `Rollback` records. And it was invented back to the days when pessimistic transactions are not supported. Now, it's unlikely to have many `Rollback` records that affect read performance.
+
+The `Lock` record is a bit more complicated. At the very beginning, when there is no pessimistic transaction or async-commit transaction, it is only used to check read-write conflicts, mostly for write-skew prevention. In pessimistic transactions, if a key is locked but not changed in the end, the pessimistic lock will be finally turned into a `Lock` record. In these cases, `Lock` records exist to cause write conflicts. If it happens to be the primary key of the transaction, it also marks the committed status. So, if the `Lock` record is only to cause write conflicts, it doesn't need to exist after any newer record is written. However, it is not true for the primary keys.
```

Comment:
The `Lock` record is also used to cause write conflicts, right ?
And, how to understand `write-skew prevention`?

Reply:
> The `Lock` record for optimistic transaction is also used to cause write conflicts, right ?
> And, how to understand `write-skew prevention`?

Yes. `write-skew prevention` is right the reason why people use `SELECT FOR UPDATE` in _optimistic_ transactions. For example, select where id = 1 and use the result to update the row id = 2. There can be a write-skew if we don't check conflict on id = 1.



> I executed the optimistic transaction like "begin optimistic;select * from t1 where name = "xxxxx" for update ;commit;". It also generated a `Lock` record for `write` CF. So, the `LOCK` record also as primary key of optimistic transaction ans marks the committed status, right?

Right. And that's why I say "However, it is not true for the primary keys.". In this case, the `Lock` cannot be collapsed.